### PR TITLE
GH-108 Allow for stopping the invoker in the middle of a batch

### PR DESCRIPTION
**Pull request by mbogoevici**

### CODE REVIEW

**Comment by mbogoevici** [comment]
Rebased on top of #107 so it can get the commit fixes for the tests
Adds tests and fix for #110 which manifests during the test.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java [hunk 1]
```diff
@@ -307,6 +307,21 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
 				@Override
 				public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
 					ListenerConsumer.this.assignedPartitions = partitions;
+					if (!ListenerConsumer.this.autoCommit) {
+						// Commit initial positions - while this is generally redundant
+						// it protects us from the case when another consumer starts
```
Comment (mbogoevici) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java]: Fix comment

File: spring-kafka/src/test/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainerTests.java [hunk 2]
```diff
@@ -608,4 +614,116 @@ public void handle(Exception thrownException, ConsumerRecord<?, ?> record) {
...
+		ConcurrentMessageListenerContainer<Integer, String> container2 =
+				new ConcurrentMessageListenerContainer<>(cf, containerProps);
+		final CountDownLatch latch = new CountDownLatch(8);
+		final Set<String> listenerThreadNames = Collections.synchronizedSet(new HashSet<String>());
+		List<String> receivedMessages = Collections.synchronizedList(new ArrayList<>());
+		containerProps.setMessageListener(new MessageListener<Integer, String>() {
+
+			@Override
+			public void onMessage(ConsumerRecord<Integer, String> message) {
+				System.out.println("auto: " + message + " on " + Thread.currentThread().getName());
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainerTests.java]: No `SOUT` in the `master` code. :smile: 

Adding CheckStyle rule:
````xml
<module name="Regexp">
	<property name="format" value="System.(out|err).print" />
	<property name="illegalPattern" value="true" />
	<property name="message" value="System.out or .err" />
</module>
````
on merge...

File: spring-kafka/src/test/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainerTests.java [hunk 3]
```diff
@@ -608,4 +614,116 @@ public void handle(Exception thrownException, ConsumerRecord<?, ?> record) {
...
+		template.sendDefault(0, 0, "foo");
+		template.sendDefault(0, 2, "bar");
+		template.sendDefault(0, 0, "baz");
+		template.sendDefault(0, 2, "qux");
+		template.sendDefault(1, 2, "corge");
+		template.sendDefault(1, 2, "grault");
+		template.sendDefault(1, 2, "garply");
+		template.sendDefault(1, 2, "waldo");
+		template.flush();
+		container2.start();
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainerTests.java]: Why do we need container 2 in this test?
To prove #110 ?
It isn't stopped in the end of test, BTW.
Will fix on merge if that.

File: spring-kafka/src/test/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainerTests.java (hunk 2 above)
Comment (mbogoevici) [inline spring-kafka/src/test/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainerTests.java]: oops!

File: spring-kafka/src/test/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainerTests.java [hunk 4]
```diff
@@ -608,4 +614,116 @@ public void handle(Exception thrownException, ConsumerRecord<?, ?> record) {
...
+		assertThat(latch.await(60, TimeUnit.SECONDS)).isTrue();
+		assertThat(receivedMessages).containsOnlyOnce("foo", "bar", "baz", "qux", "corge", "grault", "garply", "waldo");
+		// all messages are received
+		assertThat(receivedMessages).hasSize(8);
+		// messages are received on separate threads
+		assertThat(listenerThreadNames.size()).isGreaterThanOrEqualTo(2);
+		container.stop();
+		this.logger.info("Stop auto");
+	}
 }
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainerTests.java]: Blank line before the last class `}`. Codestyle :smile:

File: spring-kafka/src/test/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainerTests.java (hunk 3 above)
Comment (mbogoevici) [inline spring-kafka/src/test/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainerTests.java]: Yeah, it needs to be stopped. What we want is to cause a rebalance. Probably we will need some additional way to prove that a container didn't process an entire batch before stopping.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-kafka/commit/67ffc2e331672d5cb4832fba8db5fc075c19aaa6

**Pull request closed without merging**